test/tests/props.js

```diff
@@ -126,7 +126,87 @@ describe('zoid props cases', () => {
+        const expectedAccount = '123';
+
+        return wrapPromise(({ expect }) => {
+            window.__component__ = () => {
+                return zoid.create({
+                    tag:    'test-prop-input-value-passing',
+                    url:    'mock://www.child.com/base/test/windows/child/index.htm',
+                    domain: 'mock://www.child.com',
+                    props:  {
+                        account: {
+                            type:       'string',
+                            required:   true,
+                            value:      ({ props }) => {
+                                return props.account;
+                            }
+                        },
+                        passFoo: {
+                            type:     'function',
+                            required: true
+                        }
+                    }
+                });
+            };
+
+            const component = window.__component__();
+            const instance = component({
+                run: () => `
+                    window.xprops.passFoo({ account: window.xprops.account });
+                `,
+                account: expectedAccount,
+                passFoo: expect('passFoo', ({ account }) => {
+                    if (account !== expectedAccount) {
+                        throw new Error(`Expected account=${ expectedAccount } in the url, but got account=${ account }`);
+                    }
+
+                })
+            });
+
+            return instance.render(getBody());
+        });
+    });
+
+    it('should allow decorated functions to change their signature from the original value callback prop', () => {
+        return wrapPromise(({ expect, avoid }) => {
+            window.__component__ = () => {
+                return zoid.create({
+                    tag:    'changing-decorated-function-signature',
+                    url:    'mock://www.child.com/base/test/windows/child/index.htm',
+                    domain: 'mock://www.child.com',
+                    props:  {
+                        onSuccess: {
+                            type:     'function',
+                            decorate: ({ value, onError }) => {
+                                return (err, result) => {
+                                    if (err || !result) {
+                                        return onError(err);
+                                    }
+
+                                    return value(true);
+                                };
+                            }
+                        }
+                    }
+                });
+            };
+
+            const component = window.__component__();
+            const instance = component({
+                run: () => `
+                    window.xprops.onSuccess(null, {success: true});
+                `,
+                onSuccess:   expect('onSuccess'),
```

Comment:
Can we add an assertion here when `onSuccess` is called that we got the correct argument with a value of `true`?

Reply:
you got it!

Reply:
As an academic question, I wonder why Flow allowed this in the first place? Probably the answer is Flow generics not being great. Moving to TS will probably flag up a lot of weird stuff like this where the signature changes.